files/files.go

```diff
@@ -0,0 +1,153 @@
+package files
+
+import (
+	"archive/zip"
+	"bytes"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+)
+
+func ExtractAll(files []File) error {
+	for _, f := range files {
+		err := f.extract()
+		if err != nil {
+			return fmt.Errorf("Error extracing file %v: %v", f.Path, err)
+		}
+	}
+	return nil
+}
+
+// originally copied from
+// https://github.com/taskcluster/generic-worker/blob/01b581bfac21167dbd90a88ac0f81fd952cc0a64/aws.go
+
+type File struct {
+	Description string `json:"description"`
+	Path        string `json:"path"`
+	Content     string `json:"content"`
+	Encoding    string `json:"encoding"`
+	Format      string `json:"format"`
+}
+
+func (f File) extract() error {
+	switch f.Format {
+	case "file":
+		return f.extractFile()
+	case "zip":
+		return f.extractZip()
+	default:
+		return errors.New("Unknown file format " + f.Format + " in worker files")
+	}
+}
+
+func (f File) extractFile() error {
+	switch f.Encoding {
+	case "base64":
+		data, err := base64.StdEncoding.DecodeString(f.Content)
+		if err != nil {
+			return err
+		}
+		log.Printf("Writing %v to path %v", f.Description, f.Path)
+		dir := filepath.Dir(f.Path)
+		err = os.MkdirAll(dir, 0777)
+		if err != nil {
+			return err
+		}
+		return ioutil.WriteFile(f.Path, data, 0777)
+	default:
+		return errors.New("Unsupported encoding " + f.Encoding + " for worker file")
+	}
+}
+
+func (f File) extractZip() error {
+	switch f.Encoding {
+	case "base64":
+		data, err := base64.StdEncoding.DecodeString(f.Content)
+		if err != nil {
+			return err
+		}
+		log.Printf("Unzipping %v to path %v", f.Description, f.Path)
+		dir := filepath.Dir(f.Path)
+		err = os.MkdirAll(dir, 0777)
+		if err != nil {
+			return err
+		}
+		return unzip(data, f.Path)
+	default:
+		return errors.New("Unsupported encoding " + f.Encoding + " for worker file")
+	}
+}
+
+// This is a modified version of
+// http://stackoverflow.com/questions/20357223/easy-way-to-unzip-file-with-golang
+// to work with in memory zip, rather than a file
```

Comment:
LGTM, interesting that there are three separate places where `os.MkdirAll` is called, but I think the cases they cover make sense. Is there a reason the parent dir is `0755` and the subdirs of the extracted archive are `0777`?

Reply:
I'm not sure what's up with the permissions.  The "file" type adds files with 0777.  I suspect we're just relying on umask to do the right thing here?  But I can change those dirs to 0755 to match.